File: internal/license/fingerprint.go
package license

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strings"
	"time"
)

// FingerprintConfig points the collector at host-side filesystem
// mounts. Inside a container, /sys and /proc reflect the namespace,
// so we bind-mount the host's sysfs and procfs at fixed paths.
//
// Recommended docker-compose volumes:
//
//	volumes:
//	  - /sys:/host/sys:ro
//	  - /proc/cmdline:/host/cmdline:ro
//	  - /proc/driver/nvidia:/proc/driver/nvidia:ro
//
// Mounting the entire /sys (read-only) is required because sysfs uses
// internal symlinks (e.g. /sys/class/net/eth0 -> ../../devices/...)
// which only resolve correctly when the whole tree is visible inside
// the container. The verifier and hwinfo MUST share the exact same
// FingerprintConfig.
type FingerprintConfig struct {
	// DMIDir is /sys/class/dmi/id (or its host bind-mount). On ARM
	// OEM boards (e.g. XFUSION GB10) the serial fields are usually
	// empty; only sys_vendor / product_name come back populated.
	DMIDir string
	// DeviceTreeDir is /proc/device-tree (preferred) — usually present
	// on Tegra/Jetson/RPi platforms.
	DeviceTreeDir string
	// FirmwareTreeDir is /sys/firmware/devicetree/base, the sysfs
	// fallback for kernels that no longer expose /proc/device-tree
	// or systems that boot via ACPI.
	FirmwareTreeDir string
	// NetDir is /sys/class/net (root of NIC entries).
	NetDir string
	// NICName is the host NIC whose MAC participates in the fingerprint
	// (e.g. "eth0", "enP7s7"). Pinning the name prevents an attacker
	// from picking a random veth or USB adapter.
	NICName string
	// CmdlinePath is /proc/cmdline; used to extract the root= UUID.
	CmdlinePath string
	// BlockDir is /sys/class/block. The most stable identity source
	// on OEM ARM systems where DMI serials come blank — NVMe SSDs
	// always carry a factory serial in <BlockDir>/<dev>/device/serial.
	BlockDir string
	// DiskName pins a specific block device (e.g. "nvme0n1"). When
	// empty the collector picks the lexicographically-first NVMe
	// namespace, falling back to the first eMMC.
	DiskName string
	// NvidiaSMI is the path to nvidia-smi if available. Optional; the
	// GPU UUID source falls back to /proc/driver/nvidia parsing.
	NvidiaSMI string
	// RequireGPU forces failure when GPU UUID is unreadable. Set true
	// on devices that are supposed to have an NVIDIA GPU (like GB10).
	RequireGPU bool
}

// DefaultLinuxConfig returns the recommended layout for a Linux device
// where the verifier runs inside a container with /sys bind-mounted
// read-only at /host/sys. See FingerprintConfig comment for the
// docker-compose snippet.
func DefaultLinuxConfig() FingerprintConfig {
	return FingerprintConfig{
		DMIDir:          "/host/sys/class/dmi/id",
		DeviceTreeDir:   "/proc/device-tree",
		FirmwareTreeDir: "/host/sys/firmware/devicetree/base",
		NetDir:          "/host/sys/class/net",
		NICName:         firstEnv("HW_NIC", "eth0"),
		CmdlinePath:     "/host/cmdline",
		BlockDir:        "/host/sys/class/block",
		DiskName:        os.Getenv("HW_DISK"),
		NvidiaSMI:       "nvidia-smi",
		RequireGPU:      false,
	}
}

func firstEnv(name, def string) string {
	if v := os.Getenv(name); v != "" {
		return v
	}
	return def
}

// CollectSources gathers every available identity source. Sources that
// are not available are simply omitted from the map; callers decide
// whether to treat missing sources as fatal.
//
// The map keys are stable identifiers; do NOT rename them or you break
// existing licenses.
func CollectSources(cfg FingerprintConfig) (map[string]string, error) {
	sources := map[string]string{}

	// 1) x86_64 DMI / SMBIOS — every server/workstation BIOS fills these.
	//    Per-device:
	addIfNonEmpty(sources, "product_uuid", readTrimmed(filepath.Join(cfg.DMIDir, "product_uuid")))
	addIfNonEmpty(sources, "product_serial", readTrimmed(filepath.Join(cfg.DMIDir, "product_serial")))
	addIfNonEmpty(sources, "board_serial", readTrimmed(filepath.Join(cfg.DMIDir, "board_serial")))
	//    Model-level (OEM brand + product family). On ARM OEMs these
	//    are often the only populated DMI fields; they don't add
	//    per-unit entropy but they refuse to validate when the license
	//    is moved between different product lines.
	addIfNonEmpty(sources, "dmi_sys_vendor", readTrimmed(filepath.Join(cfg.DMIDir, "sys_vendor")))
	addIfNonEmpty(sources, "dmi_product_name", readTrimmed(filepath.Join(cfg.DMIDir, "product_name")))

	// 2) ARM SoC device tree — typical on Tegra Jetson / RPi. Newer
	//    kernels (and some Grace boards) drop /proc/device-tree, so we
	//    also try the sysfs path /sys/firmware/devicetree/base.
	addIfNonEmpty(sources, "dt_serial_number",
		firstNonEmpty(
			readTrimmedNullStripped(filepath.Join(cfg.DeviceTreeDir, "serial-number")),
			readTrimmedNullStripped(filepath.Join(cfg.FirmwareTreeDir, "serial-number")),
		))
	addIfNonEmpty(sources, "dt_model",
		firstNonEmpty(
			readTrimmedNullStripped(filepath.Join(cfg.DeviceTreeDir, "model")),
			readTrimmedNullStripped(filepath.Join(cfg.FirmwareTreeDir, "model")),
		))

	// 3) Pinned host NIC. We require an explicit name so attackers can't
	//    swap in another adapter. We refuse to fall back to "first NIC".
	if cfg.NICName != "" {
		if mac := readTrimmed(filepath.Join(cfg.NetDir, cfg.NICName, "address")); mac != "" && mac != "00:00:00:00:00:00" {
			sources["host_mac"] = strings.ToLower(mac)
		}
	}

	// 4) Root filesystem UUID from kernel cmdline. Stable across reboots
	//    but not across re-installs; treat as auxiliary, not primary.
	if uuid := extractRootUUID(readTrimmed(cfg.CmdlinePath)); uuid != "" {
		sources["root_uuid"] = uuid
	}

	// 5) Block-device identity. The strongest source on OEM ARM boards
	//    where DMI serials come blank — NVMe SSDs always carry a stable
	//    factory serial and (usually) an EUI-64 / NGUID wwid.
	if disk := chooseBlockDevice(cfg); disk != "" {
		addIfNonEmpty(sources, "disk_serial",
			readTrimmed(filepath.Join(cfg.BlockDir, disk, "device", "serial")))
		addIfNonEmpty(sources, "disk_wwid",
			readTrimmed(filepath.Join(cfg.BlockDir, disk, "wwid")))
	}

	// 6) GPU UUID (NVIDIA only). Two ways: nvidia-smi (preferred, works
	//    in containers with the nvidia runtime) or the proc node.
	if uuid := queryGPUUUID(cfg); uuid != "" {
		sources["gpu_uuid"] = uuid
	} else if cfg.RequireGPU {
		return nil, fmt.Errorf("gpu uuid required but unavailable")
	}

	if len(sources) == 0 {
		return nil, errors.New("no hardware identity sources available")
	}
	return sources, nil
}

// ComputeFingerprint serialises the sources map with deterministic
// ordering and returns sha256(canonical) as hex.
func ComputeFingerprint(sources map[string]string) (string, error) {
	if len(sources) == 0 {
		return "", errors.New("empty sources")
	}
	keys := make([]string, 0, len(sources))
	for k := range sources {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	h := sha256.New()
	for i, k := range keys {
		if i > 0 {
			h.Write([]byte{0x1f}) // ASCII unit-separator, never appears in our values
		}
		h.Write([]byte(k))
		h.Write([]byte{0x3d}) // '='
		h.Write([]byte(sources[k]))
	}
	return hex.EncodeToString(h.Sum(nil)), nil
}

// CollectHardwareInfo is the one-stop call used by `hwinfo` and the
// verifier alike. It returns the full HardwareInfo struct ready to be
// serialized.
func CollectHardwareInfo(cfg FingerprintConfig) (*HardwareInfo, error) {
	sources, err := CollectSources(cfg)
	if err != nil {
		return nil, err
	}
	fp, err := ComputeFingerprint(sources)
	if err != nil {
		return nil, err
	}
	return &HardwareInfo{
		SchemaVersion: 1,
		CollectedAt:   time.Now().UTC(),
		Platform:      runtime.GOOS + "/" + runtime.GOARCH,
		NIC:           cfg.NICName,
		Sources:       sources,
		Fingerprint:   fp,
	}, nil
}

// --- internals ---------------------------------------------------------

var (
	nvmeNamespaceRe = regexp.MustCompile(`^nvme\d+n\d+$`)
	mmcDeviceRe     = regexp.MustCompile(`^mmcblk\d+$`)
)

// chooseBlockDevice returns the basename of the block device whose
// `device/serial` will be hashed into the fingerprint. Selection rules:
//
//   1. If cfg.DiskName is set, use it iff its serial node is readable.
//   2. Otherwise pick the lexicographically-first nvme*n* entry.
//   3. Failing that, the first mmcblk* entry.
//
// Empty string means "no block-device source available"; the caller
// will silently omit the disk_serial / disk_wwid entries.
func chooseBlockDevice(cfg FingerprintConfig) string {
	if cfg.BlockDir == "" {
		return ""
	}
	if cfg.DiskName != "" {
		if _, err := os.Stat(filepath.Join(cfg.BlockDir, cfg.DiskName, "device", "serial")); err == nil {
			return cfg.DiskName
		}
		return ""
	}
	entries, err := os.ReadDir(cfg.BlockDir)
	if err != nil {
		return ""
	}
	var nvmes, mmcs []string
	for _, e := range entries {
		name := e.Name()
		switch {
		case nvmeNamespaceRe.MatchString(name):
			nvmes = append(nvmes, name)
		case mmcDeviceRe.MatchString(name):
			mmcs = append(mmcs, name)
		}
	}
	sort.Strings(nvmes)
	sort.Strings(mmcs)
	for _, candidate := range append(nvmes, mmcs...) {
		// Confirm the serial node is actually readable before committing.
		if _, err := os.Stat(filepath.Join(cfg.BlockDir, candidate, "device", "serial")); err == nil {
			return candidate
		}
	}
	return ""
}

func firstNonEmpty(values ...string) string {
	for _, v := range values {
		if strings.TrimSpace(v) != "" {
			return v
		}
	}
	return ""
}

func addIfNonEmpty(m map[string]string, key, val string) {
	val = strings.TrimSpace(val)
	if val == "" || val == "N/A" || val == "Not Specified" || val == "Default string" {
		return
	}
	m[key] = val
}

func readTrimmed(path string) string {
	if path == "" {
		return ""
	}
	b, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(b))
}

// readTrimmedNullStripped is for device-tree nodes which terminate
// strings with a NUL byte.
func readTrimmedNullStripped(path string) string {
	s := readTrimmed(path)
	return strings.TrimRight(s, "\x00")
}

// extractRootUUID parses kernel cmdline for `root=UUID=xxx` or
// `root=PARTUUID=xxx`. We don't accept bare device paths because they
// aren't stable identifiers.
func extractRootUUID(cmdline string) string {
	for _, tok := range strings.Fields(cmdline) {
		if !strings.HasPrefix(tok, "root=") {
			continue
		}
		v := strings.TrimPrefix(tok, "root=")
		if strings.HasPrefix(v, "UUID=") {
			return "UUID:" + strings.TrimPrefix(v, "UUID=")
		}
		if strings.HasPrefix(v, "PARTUUID=") {
			return "PARTUUID:" + strings.TrimPrefix(v, "PARTUUID=")
		}
	}
	return ""
}

func queryGPUUUID(cfg FingerprintConfig) string {
	if cfg.NvidiaSMI != "" {
		ctx, cancel := commandWithTimeout(3 * time.Second)
		defer cancel()
		out, err := exec.CommandContext(ctx, cfg.NvidiaSMI, "--query-gpu=uuid", "--format=csv,noheader").Output()
		if err == nil {
			first := strings.SplitN(strings.TrimSpace(string(out)), "\n", 2)[0]
			if first != "" {
				return strings.TrimSpace(first)
			}
		}
	}
	// Fallback: /proc/driver/nvidia/gpus/*/information contains a GPU UUID line.
	matches, _ := filepath.Glob("/proc/driver/nvidia/gpus/*/information")
	for _, p := range matches {
		data := readTrimmed(p)
		for _, line := range strings.Split(data, "\n") {
			if strings.HasPrefix(line, "GPU UUID:") {
				return strings.TrimSpace(strings.TrimPrefix(line, "GPU UUID:"))
			}
		}
	}
	return ""
}

File: verifier-node/src/fingerprint.ts
// Hardware fingerprint collector for the device side. Mirrors
// internal/license/fingerprint.go byte-for-byte.

import { createHash } from 'node:crypto';
import { execFileSync } from 'node:child_process';
import { readFileSync, readdirSync, statSync } from 'node:fs';
import * as path from 'node:path';
import * as os from 'node:os';

import type { HardwareInfo } from './types';

export interface FingerprintConfig {
  /** /sys/class/dmi/id (or host bind-mount). */
  dmiDir: string;
  /** /proc/device-tree (preferred). */
  deviceTreeDir: string;
  /** /sys/firmware/devicetree/base (sysfs fallback). */
  firmwareTreeDir?: string;
  /** /sys/class/net root. */
  netDir: string;
  /** Host NIC name whose MAC participates in the fingerprint. */
  nicName: string;
  /** /proc/cmdline or host bind-mount. */
  cmdlinePath: string;
  /** /sys/class/block (block-device sysfs root). */
  blockDir?: string;
  /** Pinned block device (e.g. "nvme0n1"); empty = auto-pick first NVMe. */
  diskName?: string;
  /** nvidia-smi binary or empty to skip. */
  nvidiaSmi?: string;
  /** Host nvidia driver proc root (container: /host/nvidia-driver). */
  nvidiaProcDir?: string;
  /** Fail if GPU UUID is unavailable. */
  requireGpu?: boolean;
}

/**
 * Bare-metal / host-namespace defaults. Use this when the Next.js
 * process runs directly on the licensed device (e.g. systemd / PM2 +
 * `npm start` on GB10) and `/sys`, `/proc` are the kernel's real
 * filesystems rather than container bind-mounts.
 *
 * Mirrors {@code internal/license/fingerprint.go::DefaultHostConfig}
 * byte-for-byte.
 */
export function defaultHostConfig(): FingerprintConfig {
  return {
    dmiDir: '/sys/class/dmi/id',
    deviceTreeDir: '/proc/device-tree',
    firmwareTreeDir: '/sys/firmware/devicetree/base',
    netDir: '/sys/class/net',
    nicName: process.env.HW_NIC ?? 'eth0',
    cmdlinePath: '/proc/cmdline',
    blockDir: '/sys/class/block',
    diskName: process.env.HW_DISK ?? '',
    nvidiaSmi: process.env.HW_NVIDIA_SMI ?? 'nvidia-smi',
    nvidiaProcDir: process.env.HW_NVIDIA_PROC ?? '/proc/driver/nvidia',
    requireGpu: false,
  };
}

/**
 * Container defaults — the Next.js process runs inside Docker with
 * /sys bind-mounted at /host/sys, /proc/cmdline at /host/cmdline,
 * and /proc/driver/nvidia at /host/nvidia-driver. See the reference
 * docker-compose.yml at the repo root.
 */
export function defaultContainerConfig(): FingerprintConfig {
  return {
    dmiDir: '/host/sys/class/dmi/id',
    deviceTreeDir: '/proc/device-tree',
    firmwareTreeDir: '/host/sys/firmware/devicetree/base',
    netDir: '/host/sys/class/net',
    nicName: process.env.HW_NIC ?? 'eth0',
    cmdlinePath: '/host/cmdline',
    blockDir: '/host/sys/class/block',
    diskName: process.env.HW_DISK ?? '',
    nvidiaSmi: process.env.HW_NVIDIA_SMI ?? 'nvidia-smi',
    nvidiaProcDir: process.env.HW_NVIDIA_PROC ?? '/host/nvidia-driver',
    requireGpu: false,
  };
}

/**
 * Resolve the fingerprint layout the same way the Go CLI does:
 *
 *   - if `forceContainer === true` (or env `HW_CONTAINER=1`), use the
 *     container bind-mount paths (`/host/sys/...`, `/host/cmdline`)
 *   - otherwise use bare-metal host paths (`/sys/...`, `/proc/cmdline`)
 *
 * Callers may overlay extra config on top of the returned object.
 *
 * `HW_DMI`, `HW_NET`, `HW_BLOCK`, `HW_CMDLINE`, `HW_DT`, `HW_FW_TREE`
 * are honoured **only in container mode**, exactly like the Go side.
 * Applying these on bare metal silently breaks fingerprint parity
 * with `issuer sign -local`, so we refuse to do it.
 */
export function fingerprintConfigFromEnv(forceContainer?: boolean): FingerprintConfig {
  const useContainer = forceContainer ?? (process.env.HW_CONTAINER === '1');
  const cfg = useContainer ? defaultContainerConfig() : defaultHostConfig();
  if (useContainer) {
    if (process.env.HW_DMI) cfg.dmiDir = process.env.HW_DMI;
    if (process.env.HW_DT) cfg.deviceTreeDir = process.env.HW_DT;
    if (process.env.HW_FW_TREE) cfg.firmwareTreeDir = process.env.HW_FW_TREE;
    if (process.env.HW_NET) cfg.netDir = process.env.HW_NET;
    if (process.env.HW_CMDLINE) cfg.cmdlinePath = process.env.HW_CMDLINE;
    if (process.env.HW_BLOCK) cfg.blockDir = process.env.HW_BLOCK;
  }
  if (process.env.HW_REQUIRE_GPU !== undefined) {
    cfg.requireGpu = process.env.HW_REQUIRE_GPU === '1' || process.env.HW_REQUIRE_GPU === 'true';
  }
  return cfg;
}

/**
 * @deprecated Use {@link fingerprintConfigFromEnv} (env-aware) or one
 * of {@link defaultHostConfig} / {@link defaultContainerConfig}
 * (explicit). Kept for backward compatibility — historically this
 * returned **container** defaults; that behaviour is preserved.
 */
export function defaultLinuxConfig(): FingerprintConfig {
  return defaultContainerConfig();
}

export function collectSources(cfg: FingerprintConfig): Record<string, string> {
  const sources: Record<string, string> = {};

  // 1) DMI / SMBIOS.
  addIfNonEmpty(sources, 'product_uuid', readTrimmed(path.join(cfg.dmiDir, 'product_uuid')));
  addIfNonEmpty(sources, 'product_serial', readTrimmed(path.join(cfg.dmiDir, 'product_serial')));
  addIfNonEmpty(sources, 'board_serial', readTrimmed(path.join(cfg.dmiDir, 'board_serial')));
  addIfNonEmpty(sources, 'dmi_sys_vendor', readTrimmed(path.join(cfg.dmiDir, 'sys_vendor')));
  addIfNonEmpty(sources, 'dmi_product_name', readTrimmed(path.join(cfg.dmiDir, 'product_name')));

  // 2) Device tree, with sysfs fallback.
  addIfNonEmpty(
    sources,
    'dt_serial_number',
    firstNonEmpty(
      readTrimmedNullStripped(path.join(cfg.deviceTreeDir, 'serial-number')),
      cfg.firmwareTreeDir ? readTrimmedNullStripped(path.join(cfg.firmwareTreeDir, 'serial-number')) : ''
    )
  );
  addIfNonEmpty(
    sources,
    'dt_model',
    firstNonEmpty(
      readTrimmedNullStripped(path.join(cfg.deviceTreeDir, 'model')),
      cfg.firmwareTreeDir ? readTrimmedNullStripped(path.join(cfg.firmwareTreeDir, 'model')) : ''
    )
  );

  // 3) Pinned NIC.
  if (cfg.nicName) {
    const mac = readTrimmed(path.join(cfg.netDir, cfg.nicName, 'address'));
    if (mac && mac !== '00:00:00:00:00:00') {
      sources['host_mac'] = mac.toLowerCase();
    }
  }

  // 4) Root UUID from kernel cmdline.
  const rootUuid = extractRootUuid(readTrimmed(cfg.cmdlinePath));
  if (rootUuid) sources['root_uuid'] = rootUuid;

  // 5) Block-device identity (the strongest source on OEM ARM boards).
  if (cfg.blockDir) {
    const disk = chooseBlockDevice(cfg);
    if (disk) {
      addIfNonEmpty(sources, 'disk_serial', readTrimmed(path.join(cfg.blockDir, disk, 'device', 'serial')));
      addIfNonEmpty(sources, 'disk_wwid', readTrimmed(path.join(cfg.blockDir, disk, 'wwid')));
    }
  }

  // 6) GPU UUID.
  const gpu = queryGpuUuid(cfg);
  if (gpu) {
    sources['gpu_uuid'] = gpu;
  } else if (cfg.requireGpu) {
    throw new Error('gpu uuid required but unavailable');
  }

  if (Object.keys(sources).length === 0) {
    throw new Error('no hardware identity sources available');
  }
  return sources;
}

export function computeFingerprint(sources: Record<string, string>): string {
  const keys = Object.keys(sources).sort();
  if (keys.length === 0) throw new Error('empty sources');
  const h = createHash('sha256');
  keys.forEach((k, i) => {
    if (i > 0) h.update(Buffer.from([0x1f]));
    h.update(k, 'utf8');
    h.update(Buffer.from([0x3d]));
    h.update(sources[k]!, 'utf8');
  });
  return h.digest('hex');
}

export function collectHardwareInfo(cfg: FingerprintConfig): HardwareInfo {
  const sources = collectSources(cfg);
  const fp = computeFingerprint(sources);
  return {
    schemaVersion: 1,
    collectedAt: new Date().toISOString(),
    platform: `${os.platform()}/${os.arch()}`,
    nic: cfg.nicName,
    sources,
    fingerprint: fp,
  };
}

// --- internals --------------------------------------------------------------

function addIfNonEmpty(map: Record<string, string>, key: string, val: string): void {
  const v = val.trim();
  if (!v || v === 'N/A' || v === 'Not Specified' || v === 'Default string') return;
  map[key] = v;
}

function readTrimmed(filePath: string): string {
  if (!filePath) return '';
  try {
    statSync(filePath);
  } catch {
    return '';
  }
  try {
    return readFileSync(filePath, 'utf8').trim();
  } catch {
    return '';
  }
}

function readTrimmedNullStripped(filePath: string): string {
  return readTrimmed(filePath).replace(/\x00+$/g, '');
}

function extractRootUuid(cmdline: string): string {
  for (const tok of cmdline.split(/\s+/)) {
    if (!tok.startsWith('root=')) continue;
    const v = tok.slice(5);
    if (v.startsWith('UUID=')) return 'UUID:' + v.slice(5);
    if (v.startsWith('PARTUUID=')) return 'PARTUUID:' + v.slice(9);
  }
  return '';
}

const NVME_NAMESPACE_RE = /^nvme\d+n\d+$/;
const MMC_DEVICE_RE = /^mmcblk\d+$/;

function chooseBlockDevice(cfg: FingerprintConfig): string {
  if (!cfg.blockDir) return '';
  if (cfg.diskName) {
    try {
      statSync(path.join(cfg.blockDir, cfg.diskName, 'device', 'serial'));
      return cfg.diskName;
    } catch {
      return '';
    }
  }
  let entries: string[];
  try {
    entries = readdirSync(cfg.blockDir);
  } catch {
    return '';
  }
  const nvmes: string[] = [];
  const mmcs: string[] = [];
  for (const name of entries) {
    if (NVME_NAMESPACE_RE.test(name)) nvmes.push(name);
    else if (MMC_DEVICE_RE.test(name)) mmcs.push(name);
  }
  nvmes.sort();
  mmcs.sort();
  for (const candidate of [...nvmes, ...mmcs]) {
    try {
      statSync(path.join(cfg.blockDir, candidate, 'device', 'serial'));
      return candidate;
    } catch {
      // try next
    }
  }
  return '';
}

function firstNonEmpty(...values: string[]): string {
  for (const v of values) {
    if (v && v.trim() !== '') return v;
  }
  return '';
}

function queryGpuUuid(cfg: FingerprintConfig): string {
  if (cfg.nvidiaSmi) {
    try {
      const out = execFileSync(cfg.nvidiaSmi, ['--query-gpu=uuid', '--format=csv,noheader'], {
        encoding: 'utf8',
        timeout: 3000,
        stdio: ['ignore', 'pipe', 'ignore'],
      });
      const first = out.trim().split('\n')[0]?.trim();
      if (first) return first;
    } catch {
      // fall through
    }
  }
  try {
    const dir = path.join(cfg.nvidiaProcDir ?? '/proc/driver/nvidia', 'gpus');
    for (const entry of readdirSync(dir)) {
      const info = readTrimmed(path.join(dir, entry, 'information'));
      for (const line of info.split('\n')) {
        if (line.startsWith('GPU UUID:')) {
          return line.slice('GPU UUID:'.length).trim();
        }
      }
    }
  } catch {
    // ignore
  }
  return '';
}
